File: backend/ai_service_directo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
OaxIntegra IA — Servicio de IA en Python
=========================================
Backend mínimo con FastAPI que sirve de puente entre la página y la IA.
Mantiene la clave de API en el servidor (segura, no expuesta al navegador)
y resuelve CORS de una vez por todas.

INSTALAR:
    pip install fastapi uvicorn httpx

CONFIGURAR:
    export OPENAI_API_KEY="sk-..."       # o ANTHROPIC_API_KEY
    export PROVEEDOR="openai"            # openai | anthropic | groq | deepseek

CORRER:
    uvicorn ai_service_directo:app --host 0.0.0.0 --port 8000

USAR DESDE LA APP:
    En app.src.html, cambia:
        WEBHOOK_ASISTENTE: 'http://localhost:8000/chat'
    (o la URL pública donde lo despliegues)

DESPLEGAR GRATIS: Render, Railway, Fly.io o Deta aceptan FastAPI directo.
"""

import os
from typing import Optional

import httpx
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel

# ---------------------------------------------------------------------------
# Configuración
# ---------------------------------------------------------------------------
PROVEEDOR = os.getenv("PROVEEDOR", "openai")

CLAVES = {
    "openai":    os.getenv("OPENAI_API_KEY", ""),
    "anthropic": os.getenv("ANTHROPIC_API_KEY", ""),
    "groq":      os.getenv("GROQ_API_KEY", ""),
    "deepseek":  os.getenv("DEEPSEEK_API_KEY", ""),
}

MODELOS = {
    "openai":    "gpt-4o-mini",
    "anthropic": "claude-3-5-haiku-20241022",
    "groq":      "llama-3.3-70b-versatile",
    "deepseek":  "deepseek-chat",
}

ENDPOINTS = {
    "openai":    "https://api.openai.com/v1/chat/completions",
    "anthropic": "https://api.anthropic.com/v1/messages",
    "groq":      "https://api.groq.com/openai/v1/chat/completions",
    "deepseek":  "https://api.deepseek.com/chat/completions",
}

TIEMPO_LIMITE = 25.0


# ---------------------------------------------------------------------------
# Prompt maestro — la personalidad de OaxIntegra IA
# ---------------------------------------------------------------------------
def prompt_maestro(giro: str = "", usuario: str = "") -> str:
    giro = giro or "un negocio pequeño"
    usuario = usuario or "el emprendedor"
    return f"""Eres el asistente de OaxIntegra IA, una plataforma hecha para emprendedores
oaxaqueños tradicionales: artesanos, mezcaleros, cocineras, comerciantes y
gente del turismo en Oaxaca, México.

CON QUIÉN HABLAS:
- Se llama {usuario} y su negocio es de: {giro}.
- Puede tener poca práctica con la tecnología y algo de desconfianza hacia ella.
- Valora profundamente su cultura y su oficio.

CÓMO DEBES HABLAR:
- Español mexicano cálido, cercano y respetuoso. Como un ayudante de confianza.
- PROHIBIDO usar tecnicismos: nunca digas "prompt", "token", "IA generativa",
  "endpoint", "algoritmo", "input". Di "escríbeme", "tu mensaje", "la máquina".
- Frases cortas. Nada de párrafos densos.

QUÉ DEBES ENTREGAR:
- SIEMPRE material listo para usar, no consejos vagos.
- Si pide una publicación: escríbela completa, lista para copiar y pegar.
- Si pregunta por precios: da un rango concreto, explicado en lenguaje simple.
- Si pide una descripción de producto: escríbela con el habla de la región.
- Usa ejemplos de SU giro ({giro}), no ejemplos genéricos.

RESPETO CULTURAL:
- La IA acompaña y ayuda, NUNCA reemplaza el trabajo artesanal.
- Valora el oficio tradicional. Nunca sugieras industrializar ni abaratar
  lo hecho a mano.

LARGO: máximo 200 palabras, salvo que pidan explícitamente más."""


# ---------------------------------------------------------------------------
# App
# ---------------------------------------------------------------------------
app = FastAPI(title="OaxIntegra IA — servicio de IA")

# CORS abierto: esto resuelve el problema que n8n no resolvía
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=False,
    allow_methods=["POST", "OPTIONS", "GET"],
    allow_headers=["*"],
)


class Peticion(BaseModel):
    message: str
    username: Optional[str] = "invitado"
    code: Optional[str] = ""
    businessType: Optional[str] = ""
    phone: Optional[str] = "No proporcionado"
    origen: Optional[str] = "oaxintegra"


def _armar(consulta: str, sistema: str):
    """Devuelve (url, headers, body) según el proveedor activo."""
    clave = CLAVES.get(PROVEEDOR, "")
    modelo = MODELOS[PROVEEDOR]
    url = ENDPOINTS[PROVEEDOR]

    if PROVEEDOR in ("openai", "groq", "deepseek"):
        return url, {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {clave}",
        }, {
            "model": modelo,
            "max_tokens": 900,
            "messages": [
                {"role": "system", "content": sistema},
                {"role": "user", "content": consulta},
            ],
        }

    if PROVEEDOR == "anthropic":
        return url, {
            "Content-Type": "application/json",
            "x-api-key": clave,
            "anthropic-version": "2023-06-01",
        }, {
            "model": modelo,
            "max_tokens": 900,
            "system": sistema,
            "messages": [{"role": "user", "content": consulta}],
        }

    raise ValueError(f"Proveedor no reconocido: {PROVEEDOR}")


def _extraer(d: dict) -> str:
    """Extractor universal del texto, sirva el proveedor que sirva."""
    if not d:
        return ""
    if isinstance(d, str):
        return d
    # OpenAI / Groq / DeepSeek
    try:
        return d["choices"][0]["message"]["content"]
    except (KeyError, IndexError, TypeError):
        pass
    # Anthropic
    try:
        return "".join(b["text"] for b in d["content"] if b.get("type") == "text")
    except (KeyError, TypeError):
        pass
    # genéricos
    for k in ("output", "respuesta", "reply", "text", "message", "answer", "result"):
        v = d.get(k)
        if isinstance(v, str) and v.strip():
            return v
    return ""


@app.get("/")
def salud():
    return {"ok": True, "proveedor": PROVEEDOR, "modelo": MODELOS.get(PROVEEDOR)}


@app.post("/chat")
async def chat(p: Peticion):
    """
    Recibe exactamente el mismo payload que la app mandaba a n8n,
    así no hay que cambiar nada en el frontend salvo la URL.
    """
    sistema = prompt_maestro(p.businessType, p.username)
    url, headers, body = _armar(p.message, sistema)

    try:
        async with httpx.AsyncClient(timeout=TIEMPO_LIMITE) as cliente:
            r = await cliente.post(url, headers=headers, json=body)
            r.raise_for_status()
            texto = _extraer(r.json())
        return {"output": texto}
    except httpx.TimeoutException:
        return {"output": "", "error": "La IA tardó demasiado en responder."}
    except Exception as e:  # noqa: BLE001
        return {"output": "", "error": f"No pude contactar a la IA: {e}"}


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)

File: backend/ai_service_directo.js
/* ============================================================================
   OaxIntegra IA — Servicio de IA DIRECTO (sin n8n)
   ----------------------------------------------------------------------------
   Reemplaza la llamada al webhook por una llamada directa al proveedor de IA.
   Elimina la capa de n8n, que es la que ha estado fallando (CORS / flujo
   inactivo).

   CÓMO USARLO EN app.src.html:
   1. Copia este archivo dentro del <script> de la app (o inclúyelo aparte)
   2. En CONFIG, pon tu clave y el proveedor
   3. Sustituye la llamada `enviarWebhook(CONFIG.WEBHOOK_ASISTENTE, {...})`
      por `pedirALaIA(consulta, sesion)`

   ⚠️ SEGURIDAD: poner la clave aquí la expone a cualquiera que vea el código
   fuente. Aceptable para prueba personal. Para producción usa el proxy
   serverless del final de este archivo.
   ========================================================================= */

var CONFIG_IA = {
  /* Elige uno: 'openai' | 'anthropic' | 'groq' | 'deepseek' | 'gemini' */
  PROVEEDOR: 'openai',

  CLAVES: {
    openai:    'sk-TU_CLAVE_AQUI',
    anthropic: 'sk-ant-TU_CLAVE_AQUI',
    groq:      'gsk_TU_CLAVE_AQUI',
    deepseek:  'sk-TU_CLAVE_AQUI',
    gemini:    'TU_CLAVE_AQUI'
  },

  MODELOS: {
    openai:    'gpt-4o-mini',
    anthropic: 'claude-3-5-haiku-20241022',
    groq:      'llama-3.3-70b-versatile',
    deepseek:  'deepseek-chat',
    gemini:    'gemini-2.0-flash'
  },

  LIMITE_MS: 20000,

  /* Si prefieres un proxy propio (recomendado en producción), pon aquí su URL
     y se usará en lugar de llamar al proveedor directamente. */
  PROXY: ''   /* ej: 'https://tu-sitio.netlify.app/.netlify/functions/ia' */
};


/* ---------------------------------------------------------------------------
   PROMPT MAESTRO — el alma de OaxIntegra IA
   Este texto define la personalidad. Cuídalo: aquí vive el tono del proyecto.
   ------------------------------------------------------------------------ */
function promptMaestro(sesion) {
  var giro = (sesion && sesion.giroTexto) ? sesion.giroTexto : 'un negocio pequeño';
  var quien = (sesion && sesion.usuario) ? sesion.usuario : 'el emprendedor';

  return [
    'Eres el asistente de OaxIntegra IA, una plataforma hecha para emprendedores',
    'oaxaqueños tradicionales: artesanos, mezcaleros, cocineras, comerciantes y',
    'gente del turismo en Oaxaca, México.',
    '',
    'CON QUIÉN HABLAS:',
    '- Se llama ' + quien + ' y su negocio es de: ' + giro + '.',
    '- Puede tener poca práctica con la tecnología y algo de desconfianza hacia ella.',
    '- Valora profundamente su cultura y su oficio.',
    '',
    'CÓMO DEBES HABLAR:',
    '- Español mexicano cálido, cercano y respetuoso. Como un ayudante de confianza.',
    '- PROHIBIDO usar tecnicismos: nunca digas "prompt", "token", "IA generativa",',
    '  "endpoint", "algoritmo", "input". Di "escríbeme", "tu mensaje", "la máquina".',
    '- Frases cortas. Nada de párrafos densos.',
    '',
    'QUÉ DEBES ENTREGAR:',
    '- SIEMPRE material listo para usar, no consejos vagos.',
    '- Si pide una publicación: escríbela completa, lista para copiar y pegar.',
    '- Si pregunta por precios: da un rango concreto con su razonamiento en',
    '  lenguaje simple.',
    '- Si pide una descripción de producto: escríbela con el habla de la región.',
    '- Usa ejemplos de SU giro (' + giro + '), no ejemplos genéricos.',
    '',
    'RESPETO CULTURAL:',
    '- La IA acompaña y ayuda, NUNCA reemplaza el trabajo artesanal.',
    '- Valora el oficio tradicional. Nunca sugieras industrializar ni abaratar',
    '  lo hecho a mano.',
    '',
    'LARGO: máximo 200 palabras, salvo que pidan explícitamente más.'
  ].join('\n');
}


/* ---------------------------------------------------------------------------
   LLAMADA PRINCIPAL — úsala desde la app
   Devuelve una Promesa con el texto, o null si falla (→ motor local)
   ------------------------------------------------------------------------ */
function pedirALaIA(consulta, sesion) {
  var ctrl = typeof AbortController !== 'undefined' ? new AbortController() : null;
  var corte = setTimeout(function () { if (ctrl) { ctrl.abort(); } }, CONFIG_IA.LIMITE_MS);

  var peticion = CONFIG_IA.PROXY
    ? armarProxy(consulta, sesion)
    : armarProveedor(consulta, sesion);

  var opc = {
    method: 'POST',
    headers: peticion.headers,
    body: JSON.stringify(peticion.body)
  };
  if (ctrl) { opc.signal = ctrl.signal; }

  return fetch(peticion.url, opc)
    .then(function (r) {
      clearTimeout(corte);
      if (!r.ok) { throw new Error('HTTP ' + r.status); }
      return r.json();
    })
    .then(function (d) { return extraerTextoIA(d); })
    .catch(function (err) {
      clearTimeout(corte);
      if (err && err.name === 'AbortError') {
        console.info('[OaxIntegra] La IA tardó demasiado; uso modo local.');
      } else {
        console.info('[OaxIntegra] Sin respuesta de la IA:', err.message);
      }
      return null;
    });
}


/* --- armado de la petición según el proveedor --------------------------- */
function armarProveedor(consulta, sesion) {
  var p = CONFIG_IA.PROVEEDOR;
  var clave = CONFIG_IA.CLAVES[p];
  var modelo = CONFIG_IA.MODELOS[p];
  var sistema = promptMaestro(sesion);

  /* OpenAI, Groq y DeepSeek comparten el formato de OpenAI */
  if (p === 'openai' || p === 'groq' || p === 'deepseek') {
    var urls = {
      openai:   'https://api.openai.com/v1/chat/completions',
      groq:     'https://api.groq.com/openai/v1/chat/completions',
      deepseek: 'https://api.deepseek.com/chat/completions'
    };
    return {
      url: urls[p],
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ' + clave
      },
      body: {
        model: modelo,
        max_tokens: 900,
        messages: [
          { role: 'system', content: sistema },
          { role: 'user',   content: consulta }
        ]
      }
    };
  }

  if (p === 'anthropic') {
    return {
      url: 'https://api.anthropic.com/v1/messages',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': clave,
        'anthropic-version': '2023-06-01',
        'anthropic-dangerous-direct-browser-access': 'true'
      },
      body: {
        model: modelo,
        max_tokens: 900,
        system: sistema,
        messages: [{ role: 'user', content: consulta }]
      }
    };
  }

  if (p === 'gemini') {
    return {
      url: 'https://generativelanguage.googleapis.com/v1beta/models/'
           + modelo + ':generateContent?key=' + clave,
      headers: { 'Content-Type': 'application/json' },
      body: {
        system_instruction: { parts: [{ text: sistema }] },
        contents: [{ parts: [{ text: consulta }] }]
      }
    };
  }

  throw new Error('Proveedor no reconocido: ' + p);
}

/* --- si usas tu propio proxy serverless --------------------------------- */
function armarProxy(consulta, sesion) {
  return {
    url: CONFIG_IA.PROXY,
    headers: { 'Content-Type': 'application/json' },
    body: {
      message: consulta,
      username: sesion ? sesion.usuario : 'invitado',
      businessType: sesion ? sesion.giroTexto : '',
      system: promptMaestro(sesion)
    }
  };
}


/* --- extractor universal: sirve para cualquier proveedor ---------------- */
function extraerTextoIA(d) {
  if (!d) { return ''; }
  if (typeof d === 'string') { return d; }

  /* OpenAI / Groq / DeepSeek */
  if (d.choices && d.choices[0] && d.choices[0].message) {
    return d.choices[0].message.content || '';
  }
  /* Anthropic */
  if (d.content && d.content.length) {
    var t = '';
    for (var i = 0; i < d.content.length; i++) {
      if (d.content[i].type === 'text') { t += d.content[i].text; }
    }
    if (t) { return t; }
  }
  /* Gemini */
  if (d.candidates && d.candidates[0] && d.candidates[0].content) {
    var partes = d.candidates[0].content.parts || [];
    if (partes.length) { return partes[0].text || ''; }
  }
  /* proxy propio u otros formatos */
  var llaves = ['output','respuesta','reply','text','message','answer','content','result','data'];
  for (var j = 0; j < llaves.length; j++) {
    var v = d[llaves[j]];
    if (typeof v === 'string' && v.trim()) { return v; }
    if (v && typeof v === 'object') { var an = extraerTextoIA(v); if (an) { return an; } }
  }
  return '';
}


/* ============================================================================
   PROXY SERVERLESS RECOMENDADO PARA PRODUCCIÓN
   ----------------------------------------------------------------------------
   Guarda esto como netlify/functions/ia.js y despliega en Netlify.
   La clave queda en las variables de entorno del servidor, nunca en el navegador.

   exports.handler = async (event) => {
     const headers = {
       'Access-Control-Allow-Origin': '*',
       'Access-Control-Allow-Headers': 'Content-Type',
       'Access-Control-Allow-Methods': 'POST, OPTIONS'
     };
     if (event.httpMethod === 'OPTIONS') return { statusCode: 204, headers };

     const { message, system } = JSON.parse(event.body || '{}');

     const r = await fetch('https://api.openai.com/v1/chat/completions', {
       method: 'POST',
       headers: {
         'Content-Type': 'application/json',
         'Authorization': 'Bearer ' + process.env.OPENAI_API_KEY
       },
       body: JSON.stringify({
         model: 'gpt-4o-mini',
         max_tokens: 900,
         messages: [
           { role: 'system', content: system },
           { role: 'user', content: message }
         ]
       })
     });
     const d = await r.json();
     return {
       statusCode: 200,
       headers: { ...headers, 'Content-Type': 'application/json' },
       body: JSON.stringify({ output: d.choices?.[0]?.message?.content || '' })
     };
   };
   ========================================================================= */
